File: ros_lab/scripts/lidar_processing.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Float64
from sensor_msgs.msg import LaserScan
from ros_lab.msg import ScanRange

class LidarProcessing:
    def __init__(self):

        rospy.init_node("lidar_processor_node", anonymous=True)

        # Initialize the publishers
        self.cp_pub = rospy.Publisher("/closest_point", Float64, queue_size=10)
        self.fp_pub = rospy.Publisher("/farthest_point", Float64, queue_size=10)
        self.scan_pub = rospy.Publisher("/scan_range", ScanRange, queue_size=10)

        # Initialize the subscriber
        rospy.Subscriber("scan", LaserScan, self.lidar_cb)

    def lidar_cb(self, msg):

        ranges = msg.ranges
        closest_point = min(ranges)
        farthest_point = max(ranges)

        cp = Float64()
        cp.data = closest_point

        fp = Float64()
        fp.data = farthest_point

        self.cp_pub.publish(cp)
        self.fp_pub.publish(fp)

        scan_msg = ScanRange()
        scan_msg.header = msg.header
        scan_msg.closest_point = closest_point
        scan_msg.farthest_point = farthest_point

        self.scan_pub.publish(scan_msg)

if __name__ == "__main__":
    LP = LidarProcessing()
    rospy.spin()

File: ros_lab/src/lidar_processing.cpp
#include <vector>
#include <algorithm>
#include <iostream>
#include <typeinfo>
#include <ros_lab/lidar_processing.h>
#include <ros_lab/ScanRange.h>

LidarProcessing::LidarProcessing(ros::NodeHandle &nh){
    nh_ = nh;
}

LidarProcessing::~LidarProcessing(){}

void LidarProcessing::initialize(){
    
    closest_point_pub_ = nh_.advertise<std_msgs::Float64>("/closest_point", 1);
    farthest_point_pub_ = nh_.advertise<std_msgs::Float64>("/farthest_point", 1);
    scan_msg_pub_ = nh_.advertise<ros_lab::ScanRange>("/lidar_range", 10);
    
    scan_sub_ = nh_.subscribe("/scan", 1, &LidarProcessing::laserScanCallback, this);
}

void LidarProcessing::laserScanCallback(const sensor_msgs::LaserScan::ConstPtr& scan_msg){
    
    std::vector<float> ranges;
    std_msgs::Float64 closest_point;
    std_msgs::Float64 farthest_point;

    ros_lab::ScanRange lidar_msg;

    ranges = scan_msg->ranges;
    closest_point.data = *std::min_element(ranges.begin(), ranges.end());
    farthest_point.data = *std::max_element(ranges.begin(), ranges.end());

    closest_point_pub_.publish(closest_point);
    farthest_point_pub_.publish(farthest_point);

    lidar_msg.header = scan_msg->header;
    lidar_msg.closest_point = closest_point.data;
    lidar_msg.farthest_point = farthest_point.data;

    scan_msg_pub_.publish(lidar_msg);
}
